pkg/archive/archive.go

```
@@ -856,179 +856,189 @@ func Tar(path string, compression Compression) (io.ReadCloser, error) {
 	return TarWithOptions(path, &TarOptions{Compression: compression})
 }
 
-// TarWithOptions creates an archive from the directory at `path`, only including files whose relative
-// paths are included in `options.IncludeFiles` (if non-nil) or not in `options.ExcludePatterns`.
-func TarWithOptions(srcPath string, options *TarOptions) (io.ReadCloser, error) {
+// TarWithOptionsTo writes `srcPath` as an uncompressed tar archive to
+// `dest`, only including files whose relative paths are included in
+// `options.IncludeFiles` (if non-nil) or not in
+// `options.ExcludePatterns`.
+func TarWithOptionsTo(dest io.WriteCloser, srcPath string, options *TarOptions) error {
```

Given how this function silently ignores many errors (and, it seems, the plan is to _change that_), I’m very uncomfortable with making it a public API at this point.

---

I think the “uncontested parts” can be merged ~immediately. 

I didn’t look at the failures triggered by adding error handling ~at all yet. I understand that they matter to you (don’t  they?), so they block the overall effort _anyway_. It think that’s the next step.

And making the API public would follow, or be a part of the “add error handling” PR.

---

> I think it would be better to export something like `archive.tarWriter`.

I have ~no opinion on that in principle, I’d defer to actual c/storage maintainers.

At a high level, I’m a bit unclear on what is the point of a granular API for external consumers. `archive.Tar`, sure, a very simple one-shot call. `archive.TarWithOptions`, fine, for various tunables (although many would be pretty specific to needs of c/storage callers).

For the individual `tarWriter.addFile`… I’m tempted to question why a caller specialized _enough

---

I think the “uncontested parts” can be merged ~immediately. 

I didn’t look at the failures triggered by adding error handling ~at all yet. I understand that they matter to you (don’t  they?), so they block the overall effort _anyway_. It think that’s the next step.

And making the API public would follow, or be a part of the “add error handling” PR.

---

> I think it would be better to export something like `archive.tarWriter`.

I have ~no opinion on that in principle, I’d defer to actual c/storage maintainers.

At a high level, I’m a bit unclear on what is the point of a granular API for external consumers. `archive.Tar`, sure, a very simple one-shot call. `archive.TarWithOptions`, fine, for various tunables (although many would be pretty specific to needs of c/storage callers).

For the individual `tarWriter.addFile`… I’m tempted to question whether there are callers specialized _enough_ that using just `archive.TarWithOptions` is insufficient, but not so much that they would want to use the standard-library `archive/tar.Writer`, and whether the intersection of these conditions is enough to worry about.

---

> Then, some of the functionality of TarOptions could be expressed in code by callers

I don’t know what this means. Do you mean some other parameter-passing convention (e.g. “functional options”?) Using ordinary function parameters instead of a struct? I’d be rather opposed to the latter.

---

I am primarily interested in creating tar files (in the Podman Checkpoint command) without the intermediate buffering introduced by `io.Pipe`. I’m curious, does that have a measurable performance impact?

---

> I understand that they matter to you (don’t they?), 

The failures in the test suite (caused by ignoring errors) don't matter to me. What matters to me is that in case of _any_ error, the tar creation fails.

> they would want to use the standard-library archive/tar.Writer, and whether the intersection of these conditions is enough to worry about.

The `archive/tar` package falls short because it doesn't  connect the file system to the Writer/Reader, handling among others 

* extended attributes
* file types (mkdir vs mknod vs creat)
* hard-links

Note that the std library `archive/tar.Writer` is structured as a type with methods, offering greater flexibility to callers, rather than a function with a zillion options. 

> I’m curious, does that have a measurable performance impact?

See https://github.com/containers/podman/pull/25050/commits/40e49cce65464643aeedc83a8ba0e12c1f4065bd and https://github.com/containers/podman/issues/24826#issuecomment-2601996867

for writing large files to BTRFS, there is an effective speedup of up to 1000x.